packages/Button/src/Button.tsx

```diff
@@ -267,61 +281,124 @@ const baseStyle = css`
   }
 `;
 
-export default class Button extends Component {
-  static displayName = 'Button';
+interface SharedButtonProps {
+  variant: Variant;
+  size: Size;
+  className?: string;
+  children?: React.ReactNode;
+  disabled?: boolean;
+}
 
-  static propTypes = {
-    variant: PropTypes.oneOf(['default', 'primary', 'info', 'danger', 'dark']),
-    size: PropTypes.oneOf(['xsmall', 'small', 'normal', 'large']),
-    className: PropTypes.string,
-    children: PropTypes.node,
-    disabled: PropTypes.bool,
-    as: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
-    href: PropTypes.string,
-  };
+interface LinkButtonProps extends HTMLElementProps<'a'>, SharedButtonProps {
+  href: string;
+}
+
+interface ButtonButtonProps
+  extends HTMLElementProps<'button'>,
+    SharedButtonProps {
+  href?: null;
+}
+
+type CustomElementButtonProps = SharedButtonProps & {
+  as: React.ElementType<any>;
+  [key: string]: any;
+};
+
+type ButtonProps =
+  | LinkButtonProps
+  | ButtonButtonProps
+  | CustomElementButtonProps;
+
+function hasCustomRoot(props: ButtonProps): props is CustomElementButtonProps {
```

Comment:
Little confused by this naming. What do you mean by root? Is it just `isCustomElement` ?

Reply:
The existing code had called the component variable `Root` so I was just going off of that. `usesCustomElement` could maybe be a good name? `isCustomElement` would suggest to me that the argument passed into it is being checked, when instead it's the `props` object.